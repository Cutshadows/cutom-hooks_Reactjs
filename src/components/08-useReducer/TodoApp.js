import React, { useEffect, useReducer } from 'react';
import {todoReducer} from './todoReducer';
import { useForm } from '../../hooks/useForm';

import './styles.scss';

const init=()=>{
    // return localStorage.getItem('todos')?JSON.parse(localStorage.getItem('todos')):[];
    return JSON.parse(localStorage.getItem('todos'))||[];
    // return [{
    //     id: new Date().getTime(),
    //     desc:'Aprender react',
    //     done: false
    // }];
};

const TodoApp = () => {
 
    
    const [todos, dispatch] = useReducer(todoReducer, [], init)
    const [{description}, handleInputChange, reset]=useForm({
        description:''
    });

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos))
    }, [todos])

    const handleSubmit=(e)=>{
        e.preventDefault();

        if(description.trim().length<= 1){
            return;
        }
        const newTodo={
            id: new Date().getTime(),
            desc:description,
            done: false
        };
        const action={
            type:'add',
            payload:newTodo
        }
        dispatch(action);
        reset();


    }
    return ( 
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, indice)=>(
                                <li
                                key={todo.id}
                                className="list-group-item"
                                >
                                    <p className="text-center">{indice+1}-{todo.desc}</p>
                                    <button
                                    className="btn btn-danger"
                                    >Borrar</button>
                                </li>
                            ))
                        
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar Todo</h4>
                    <hr/>
                    <form onSubmit={handleSubmit}>
                        <input type="text"
                            name="description"
                            value={description}
                            onChange={handleInputChange}
                            className="form-control"
                            placeholder="Aprender..."
                            autoComplete="off"/>
                        <button
                        type="submit"
                        className="btn btn-outline-success mt-1 w-100">Agregar</button>

                    </form>

                </div>
            </div>
           
        </div>
     );
}
 
export default TodoApp;