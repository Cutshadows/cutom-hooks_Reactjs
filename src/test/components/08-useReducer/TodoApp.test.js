const { shallow, mount } = require("enzyme")
import { act } from '@testing-library/react';
import TodoApp from '../../../components/08-useReducer/TodoApp';
import { demotTodos } from '../../fixtures/demoTodos';

describe('Pruebas sobre <TodoApp></TodoApp>', () => {


    const wrapper=shallow(<TodoApp />)

    Storage.prototype.setItem=jest.fn(()=>{});
    test('deberia de cargar todo el wrapper ', () => {
        expect(wrapper).toMatchSnapshot();
    });
    

    test('debe de agregar un Todo', () => {
        const wrapper=mount(<TodoApp />);
        act(()=>{
            wrapper.find('Formulario').prop('handleAddTodo')(demotTodos[0]);
            wrapper.find('Formulario').prop('handleAddTodo')(demotTodos[1]);
        })

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
        // expect(localStorage.setItem).toHaveBeenCalledWith({});

        
    });
    test('debe eliminar un todo', () => {
        
        wrapper.find('Formulario').prop('handleAddTodo')(demotTodos[0]);
        wrapper.find('Listado').prop('handleDelete')(demotTodos[0].id);
        // expect(wrapper.find('h1').text().trim()).toBe('TodoApp (1)');
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)');



    })
    
    
})
