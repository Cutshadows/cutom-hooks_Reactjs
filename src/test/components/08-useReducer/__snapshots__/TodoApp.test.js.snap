// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <TodoApp></TodoApp> deberia de cargar todo el wrapper  1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <Listado
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <h4>
        Agregar Todo
      </h4>
      <hr />
      <Formulario
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
